Property level security

OliviaYtterbrink opened the pull request:
Added ability to blacklist property reading for specified roles.

Currently only properties on nodes are filtered due to relationship not using the `PropertyCursor` yet.

Also clears up intent of `SecurityContext` by splitting it into `LoginContext` which handles authentication and a `SecurityContext` that holds the authorization of a user.

henriknyman commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala:
```diff
@@ -456,7 +482,13 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
+          try {
+            transactionalContext.readOperations.propertyKeyGetName(propertyKeyId)
+            buffer.append(propertyKeyId)
+          } catch {
```
Using exceptions for this seems a bit expensive. Since this need to be handled during operation in normal conditions it would be great if we could use an augmented cursor or at least use a return value to communicate the 'not found' status (we could add a new API method for this if necessary).

fickludd replied:
+1 on not using exceptions for this. Modifying the PropertyCursor would solve it :).

henriknyman commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala:
```diff
@@ -1011,7 +1049,11 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
-      if (nodeCursor.next()) fromNodeProxy(entityAccessor.newNodeProxy(nodeCursor.nodeReference()))
+      if (nodeCursor.next()) {
+        val nodeProxy = entityAccessor.newNodeProxy(nodeCursor.nodeReference())
+        val secureNodeProxy = entityAccessor.newControlledNodeProxy(nodeProxy, transactionalContext.securityContext.mode())
```
Since this is executed a lot (on all node scan) I think using inheritance instead of composition for ControlledNodeProxy may be warranted for performance reasons.

henriknyman reviewed (commented):
Some comments from a performance perspective

fickludd commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala:
```diff
@@ -196,9 +197,30 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
+    }
+    indexSearchMonitor.indexSeek(filteredIndex, values)
+    val predicates = filteredIndex.properties.zip(values).map(p => IndexQuery.exact(p._1, p._2))
+    seek(filteredIndex, predicates:_*)
```
Did you think about putting the access control in the kernel instead of here? To me that would seem like the most smooth place, and it would solve both Cypher and Core API access in one place.

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java:
```diff
@@ -298,7 +301,27 @@ public Value nodeGetProperty( long nodeId, int propertyKeyId ) throws EntityNotF
+        }
+        catch ( PropertyKeyIdNotFoundKernelException e )
+        {
+            return false;
```
Again, let's avoid exceptions here. We could add a `tokenRead().propertyKeyAllowed( ... )` or `mode.allowsPropertyKey( ... )`.

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java:
```diff
@@ -298,7 +301,27 @@ public Value nodeGetProperty( long nodeId, int propertyKeyId ) throws EntityNotF
         try ( Cursor<NodeItem> node = dataRead().nodeCursorById( statement, nodeId ) )
         {
-            return dataRead().nodeGetProperty( statement, node.get(), propertyKeyId );
+            if ( propertyIsAllowed( propertyKeyId ) )
```
This check could be made before getting the node cursor. No need to spend any more time in `OperationsFacade` though... it's gone before 3.4 comes out.

fickludd commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/EnterpriseSecurityModule.java:
```diff
@@ -334,7 +337,7 @@ private static IllegalArgumentException illegalConfiguration( String message )
     }
 
-    class SecurityConfig
+    static class SecurityConfig
```
Ooh, clean-ups!

fickludd commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/MultiRealmAuthManager.java:
```diff
@@ -297,4 +304,24 @@ public void clearAuthCache()
+        {
+            return property -> true;
+        }
+    }
```
I like the concept of pre-computing the permission check function! As it happens at runtime though, I would suggest mapping from `propertyKeyId -> boolean` instead. Also assuming that the number of unique property keys is quite small, this could be a `boolean[]` where the check is just to read the offset. Or a bitmap is we were ambitious, but let's wait with that. These things are not necessary for a first draft of this functionality, but maybe having a card for optimizing this would be nice?

fickludd reviewed (changes requested):
I like many things here, but I suggest we have a go at moving the enforcement to the new Kernel API cursors. If you catch me a day next week, we can sketch out how that could work together.

Also let's avoid using exceptions as internal signalling constructs because it's so slow.

fickludd commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/plugin/PropertyLevelSecurityIT.java:
```diff
@@ -0,0 +1,514 @@
+        execute( smith, "MATCH (n) WHERE n.name = 'Andersson' RETURN n, n.alias as alias", Collections.emptyMap(), r ->
+        {
+            assertThat( r.hasNext(), equalTo( true ) );
+            assertThat( r.next().get( "alias" ), equalTo( null ) );
```
Here we should also check that `n` doesn't contain `alias`!

fickludd commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/plugin/PropertyLevelSecurityIT.java:
```diff
@@ -0,0 +1,514 @@
+        execute( morpheus, "MATCH (n) WHERE n.name = 'Andersson' RETURN n, n.alias as alias", Collections.emptyMap(), r ->
+        {
+            assertThat( r.hasNext(), equalTo( true ) );
+            assertThat( r.next().get( "alias" ), equalTo( "neo" ) );
```
...and here that that `n` does contain `alias`.

fickludd commented on enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/plugin/PropertyLevelSecurityIT.java:
```diff
@@ -0,0 +1,514 @@
+    @Test
+    public void shouldBehaveLikeDataIsMissingForRange() throws Throwable
+    {
+        execute( neo, "CREATE (n:Person {name: 'Andersson'}) ", Collections.emptyMap() ).close();
```
For all of this setup it would be nice to make a helper method, e.g. `createNode( "Person", "name", "Andersson" )`, to make it less boilerplaty.

fickludd reviewed (commented):
I really like the rich test coverage in `PropertyLevelSecurityIT`!

OliviaYtterbrink committed:
Add tests for property level access

OliviaYtterbrink committed:
Added configuration and framework for property level security

OliviaYtterbrink committed:
Split SecurityContext into LoginContext and SecurityContext

The LoginContext holds authentication information, which is needed to
begin a transaction. It does not hold authorization information however,
this is acquired just before the transaction begins, and stored in the
SecurityContext.

OliviaYtterbrink committed:
Change name of method to better reflect usage

Also cleaned up SecurityContext interface/classes

OliviaYtterbrink committed:
Add property level security on PropertyCursor

OliviaYtterbrink committed:
Filter index seek/scan

OliviaYtterbrink committed:
Add filtering on facade for properties

This will be moved over to the new api sometime, then the check should
still be respected.

OliviaYtterbrink committed:
Fixed mocking

fickludd commented on enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/MultiRealmAuthManager.java:
```diff
@@ -305,16 +308,29 @@ public void clearAuthCache()
     }
 
-    Function<String,Boolean> getPropertyPermissions( Set<String> roles )
+    Function<Integer,Boolean> getPropertyPermissions( Set<String> roles, Token token )
```
Here we could use a `IntPredicate` instead, to remove some boxing.

fickludd reviewed (approved):
Looking really good now. Minor suggestion of function specialization, which could potentially happen during that rebase =).
